asset-transfer-secured-agreement/application-gateway-typescript/src/app.ts

```diff
@@ -0,0 +1,427 @@
+            }
+        } else {
+            console.log(`*** Failed: VerifyAssetProperties, private information about asset ${assetKey} has not been verified by ${org}`);
+        }
+    } catch (e) {
+        console.log(`${RED}*** Failed: VerifyAssetProperties - ${e}${RESET}`);
+    }
+}
+
+async function agreeToBuy(assetKey:string, org:string, price:number, contract:Contract):Promise<void> {
+    try {
+        const asset_price = {
+            asset_id: assetKey,
+            price:price,
+            trade_id: now.toString()
+        };
+        console.log(`${GREEN}--> Submit Transaction: AgreeToBuy, ${assetKey} as ${org} - endorsed by ${org}${RESET}`);
+
+        await contract.submit('AgreeToBuy',{
+            arguments:[assetKey],
+            transientData: { asset_price: JSON.stringify(asset_price) }
+        });
+
+        console.log(`*** Result: committed, ${org} has agreed to buy asset ${assetKey} for 100`);
+    } catch (e) {
+        console.log(`${RED}*** Failed: AgreeToBuy - ${e}${RESET}`);
+    }
+}
+
+async function readSalePrice(assetKey:string, org:string, contract:Contract):Promise<void> {
+    try {
+        console.log(`${GREEN}--> Evaluate Transaction: GetAssetSalesPrice, - ${assetKey} from organization ${org}${RESET}`);
+        if(org===mspIdOrg2){
+            console.log(`${GREEN}* Expected to fail as ${org} has not set a sale price${RESET}`)
+        }
+
+        const resultBytes = await contract.evaluateTransaction('GetAssetSalesPrice', assetKey);
+
+        const resultString = utf8Decoder.decode(resultBytes);
+        const result = JSON.parse(resultString);
+        console.log('*** Result: GetAssetSalesPrice', result);
+    } catch (e) {
+        console.log(`${RED}*** Failed evaluateTransaction GetAssetSalesPrice: ${e}${RESET}`);
+    }
+}
+
+async function readBidPrice(assetKey:string, org:string, contract:Contract):Promise<void> {
+    try{
+        console.log(`${GREEN}--> Evaluate Transaction: GetAssetBidPrice, - ${assetKey} from organization ${org}${RESET}`);
+        if(org===mspIdOrg1){
+            console.log(`${GREEN}* Expected to fail as Org1 has not agreed to buy${RESET}`)
+        }
+
+        const resultBytes = await contract.evaluateTransaction('GetAssetBidPrice', assetKey);
+
+        const resultString = utf8Decoder.decode(resultBytes);
+        const result = JSON.parse(resultString);
+
+        console.log('*** Result: GetAssetBidPrice', result);
+    } catch (e) {
+        console.log(`${RED}*** Failed evaluateTransaction GetAssetBidPrice: ${e}${RESET}`);
+    }
+}
+
+async function transferAsset(assetKey:string, org:string,buyerOrgID:string, price:number,contract:Contract):Promise<void>{
+    try{
+        const asset_properties = {
+            object_type: 'asset_properties',
+            asset_id: assetKey,
+            color: 'blue',
+            size: 35,
+            salt: Buffer.from(now.toString()).toString('hex')
```

Comment:
accessing current time is usually problematic - as the time will change when different peers execute the transaction. Therefore this has great potential to fail... 

Unless this is only executed on one peer, or there are other mitigating factors... which if they are are worth noting in comments.

Reply:
@mbwhite Can you help me understand the significance of passing the `salt` parameter here?

The current time is being accessed at the client side and stored as a constant value now which is  being passed to the transactions .

Reply:
This is generated in the client, not the chaincode. So there is only a single value per transaction proposal, regardless of how many peers endorse or when that happens.

Since it seems to be intended to prevent the message content being guessable from the hash of the data on the main ledger, perhaps it would be better to use something like [crypto.randomBytes()](https://nodejs.org/docs/latest-v14.x/api/crypto.html#crypto_crypto_randombytes_size_callback) or [crypto.randomUUID()](https://nodejs.org/docs/latest-v14.x/api/crypto.html#crypto_crypto_randomuuid_options) rather than current date, which is much less cryptographically secure as a salt.